lib/gems/pending/util/miq_file_storage.rb

```diff
@@ -0,0 +1,217 @@
+      raise NotImplementedError, "#{self.class}##{__callee__} is not defined"
+    end
+
+    # :call-seq:
+    #   download( local_io, remote_uri )
+    #   download( nil,      remote_uri ) { |input_writer| ... }
+    #
+    # Download a file from a remote uri.
+    #
+    # In non-block form, the remote_uri is saved to the local_io.
+    #
+    # In block form, the local_io is omitted, and it is set to a PTY writer
+    # path that will assumed to be read by the block provided.
+    def download(local_file, remote_file_uri, &block)
+      @remote_file_path = remote_file_uri
+      if block_given?
+        thread = handle_io_block(&block)
+        download_single(remote_file_uri, input_writer)
+        input_writer.close
+        thread.join
+      else
+        download_single(remote_file_uri, local_file)
+      end
+    ensure
+      reset_vars
+    end
+
+    private
+
+    def upload_single(file) # rubocop:disable Lint/UnusedMethodArgument
+      raise NotImplementedError, "#{self.class}#upload_single is not defined"
+    end
+
+    def upload_splits
+      @position = 0
+      begin
+        until source_input.eof?
+          upload_single(next_split_filename)
+          @position += byte_count
+        end
+      rescue Errno::EIO
+        # Error::EIO is raised when you read from a closed PTY on Linux.  See:
+        #
+        #   https://stackoverflow.com/a/10306782/3574689
+        #   https://ruby-doc.org/stdlib-2.2.3/libdoc/pty/rdoc/PTY.html#module-PTY-label-Example
+        #
+        # Or for the neckbeards:
+        #
+        #   $ ri PTY
+        #
+        # Done with reading at this point, but update @position... just incase
+        @position += byte_count
+      end
+    end
+
+    def initialize_upload_vars(*upload_args)
+      upload_args.pop if (@byte_count = parse_byte_value(upload_args.last))
+      @remote_file_path = upload_args.pop
+
+      unless upload_args.empty?
+        source        = upload_args.pop
+        @source_input = source.kind_of?(IO) ? source : File.open(source, "r")
+      end
+    end
+
+    def parse_byte_value(bytes)
+      match = bytes.to_s.match(BYTE_HASH_MATCH) || return
+
+      bytes = match[:BYTE_NUM].to_f
+      if match[:BYTE_QUALIFIER]
+        bytes *= BYTE_HASH[match[:BYTE_QUALIFIER].downcase]
+      end
+      bytes.to_i
+    end
+
+    def handle_io_block
+      if block_given?
+        require "tmpdir"
+
+        # create pathname, but don't create the file for it (next line)
+        fifo_path = Pathname.new(Dir::Tmpname.create("") {})
+        File.mkfifo(fifo_path)
+
+        # For #Reasons(TM), the reader must be opened first
+        @source_input = File.open(fifo_path.to_s, IO::RDONLY | IO::NONBLOCK)
+        @input_writer = File.open(fifo_path.to_s, IO::WRONLY | IO::NONBLOCK)
+
+        Thread.new do
+          yield fifo_path      # send the path to the block to get executed
+          @input_writer.close  # close the file so we know we hit EOF (for #add)
+        end
```

Comment:
Note for myself:  We are probably going to want to do something like:

```ruby
Thread.new do
  # ...
end.tap { t.abort_on_exception = true }
```

Otherwise errors can get swallowed up in this thread and subsequent threads below it, which isn't good.

Reply:
Fixed this by adding a `being`/`ensure` in the code for the thread.  Basically the same code as above, but the first line is in the main body of the `begin`, and the `.close` is part of the `ensure`.